Library/Homebrew/tap.rb

```diff
@@ -171,17 +171,34 @@ def core_tap?
   # @option options [Boolean] :quiet If set, suppress all output.
   def install(options = {})
     require "descriptions"
-    raise TapAlreadyTappedError, name if installed?
-    clear_cache
 
+    installed = installed?
+    full_clone = options.fetch(:full_clone, true)
     quiet = options.fetch(:quiet, false)
 
+    raise TapAlreadyTappedError, name if installed && !full_clone
+
+    if installed && !(path/".git/shallow").exist?
+      ohai "#{name} is already a full clone" unless quiet
+      return
+    end
+
+    clear_cache unless installed
+
     # ensure git is installed
     Utils.ensure_git_installed!
-    ohai "Tapping #{name}" unless quiet
+    if installed
+      ohai "Unshallowing #{name}" unless quiet
+    else
+      ohai "Tapping #{name}" unless quiet
+    end
     remote = options[:clone_target] || "https://github.com/#{user}/homebrew-#{repo}"
-    args = %W[clone #{remote} #{path} --config core.autocrlf=false]
-    args << "--depth=1" unless options.fetch(:full_clone, false)
+    args = if installed
+      %W[-C #{path} fetch --unshallow #{remote}]
```

Comment:
Can the remote be omitted here as it should have been properly set up during the clone? (I've never before specified a remote when calling `git fetch --unshallow`.)

Another issue might be that `-C` is not always reliable (not sure how widespread that is and which Git versions are affected) and that's why most `git` operations done by Homebrew are wrapped in something like `path.cd { <do git stuff> }` instead of using `-C`. (Sorry for being this vague. Maybe I'm also being too paranoid and all users nowadays have a sane version of Git installed.)